### ISSUE TITLE
@SpyBean does not work with generics

### ISSUE THREAD

**Issue by lincetto** [spring-projects/spring-boot#7625]
I'm trying to unit-testing a Spring batch job inside Spring boot using JUnit.
I wrote this test class where I want to spy the bean `ItemReader`:

```java
@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment=WebEnvironment.NONE)
@ActiveProfiles({"dev", "batch", "test-jobs"})
public class BatchJobTest {

	@Autowired
	private JobLauncherTestUtils jobLauncherTestUtils;
    	
	@Autowired 
	@Qualifier("contactDownloadAckJob")
	private  Job contactDownloadAckTaskJob;

	@SpyBean
	private ItemReader<CrsOscContact> reader;

	@Test
	public void testJob() throws Exception {
		given(this.reader.read()).willReturn(new CrsOscContact());
		//... blah blah blah
	}	
}
```

When I run this test, it seems that the `@SpyBean` annotation does not do its job, that should be proxying the `ItemReader` bean that's already present in the context, and so I obtain the (correct) exception because, as per definition, if the bean is not found it tries to instantiate a new bean of that type (and I have specified an interface) :

```
org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.batch.item.ItemReader]: Specified class is an interface
```

I'm pretty sure that the bean (of `ItemReader` type) is already in the context because :

 - debugging, I see that the target bean instantiation is correctly processed
 - if I change the `@SpyBean` annotation to an `@Autowired` annotation, the instance of previous point is correctly injected

**Comment by philwebb** [comment]
Could you please provide a sample application that replicates the issue?

**Comment by lincetto** [comment]
Hi @philwebb, sure!
Here a repo with a sample test application

[https://github.com/lincetto/spring-boot-ghissue-7625](https://github.com/lincetto/spring-boot-ghissue-7625)

**Comment by lincetto** [comment]
@philwebb I've got it! The problem is the use of Generics.
If I use Generics 

	@SpyBean
	private ItemReader<String> reader;

I get the exception, if I use

	@SpyBean
	private ItemReader reader;

the ApplicationContext load successfully and the bean is being spied on.
So it seems that the MockitoPostProcessor (or some sibling component) does not consider the generic type and so finish with trying to create a new instance of that type.
Core Spring instead, starting from version 4.0, consider the generic type when looking for the bean to inject.

**Comment by philwebb** [comment]
Thanks for the sample. The underlying cause does indeed seem to be the generics. I've raised [Spring JIRA SPR-15011](https://jira.spring.io/browse/SPR-15011) to see if we can improve things.

**Comment by jhoeller** [comment]
Since this is adressed in the core container in the meantime, does it work for Boot's purposes now?

**Comment by philwebb** [comment]
@jhoeller All good, I've added a test. Thanks you!

**Comment by yannanzheng** [comment]
Have you fixxed this issue? I come across it recently.

**Comment by wilkinsona** [comment]
@yannanzheng Yes, we believe the problem was fixed in 1.4.3. If you have a problem in a later version please open a new issue with a small sample that reproduces the problem.

**Comment by philwebb** [comment]
@yannanzheng Also, if you aren't using our managed dependencies, please make sure you're using at least Spring Framework 4.3.5.
